cnudie/upload.py

```diff
@@ -0,0 +1,135 @@
+import enum
+import hashlib
+
+import ccc.oci
+import gci.componentmodel as cm
+import gci.oci
+import json
+import oci.model as om
+import dataclasses
+
+
+class UploadMode(enum.Enum):
+    SKIP = 'skip'
+    FAIL = 'fail'
+    OVERWRITE = 'overwrite'
+
+
+def _normalise_component_name(component_name: str) -> str:
+    return component_name.lower()  # oci-spec allows only lowercase
+
+
+def _target_oci_ref(
+    component: cm.Component,
+    component_ref: cm.ComponentReference=None,
+    component_version: str=None,
+):
+    if not component_ref:
+        component_ref = component
+        component_name = component_ref.name
+    else:
+        component_name = component_ref.componentName
+
+    component_name = _normalise_component_name(component_name)
+    component_version = component_ref.version
+
+    last_ctx_repo = component.current_repository_ctx()
+
+    return last_ctx_repo.component_version_oci_ref(
+        name=component_name,
+        version=component_version,
+    )
+
+
+def _sha_digest(fileobj) -> str:
+    digest = hashlib.sha256()
```

Comment:
naming-nit: the object returned by hashlib.sha256 is not a digest, but a hash (function)